Deploy Automation: Automate creating and pushing a release tag

markbiek opened the pull request:
### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This updates the `yarn create-release` command with a step to create a release tag and then push the tag to `origin`.

Closes #1101.
Related to #1096.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested. Please, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

### Cancelling tagging the release
1. In the project root, run `yarn create-release`
1. Answer 'Y' to the release build prompt.
1. Answer 'N' to the "Would you like to tag this release and deploy?" prompt.
1. Verify the script stops and you are switched back to this branch.

### The tag already exists
1. Make sure the current version in `composer.json` is one that has already been released (this will be the default scenario after checking out this branch).
1. In the project root, run `yarn create-release` and answer `Y` to all prompts.
1. Once you say yes to the "Would you like to tag this release and deploy?", you'll get an `ERROR: The tag :version already exists. Deploy failed.` error.

### Fully creating and deploying a tag
1. Checkout `master`, increment the version in `composer.json`, and commit the change. This will be a fake version bump used for testing which we'll revert when we're done.
1. Checkout this branch.
1. Run `yarn create-release` and answer `Y` to all prompts.
1. At the end, you should see a `SUCCESS: Release complete!` message.
1. You should also be able to see the new tag in the [list of tags on Github](https://github.com/Automattic/wc-calypso-bridge/tags) and locally by running `git tag`.

#### Cleaning up our fake version bump
1. Run the commands `git tag -d x.x.x` and `git push --delete origin x.x.x` where `x.x.x` is the fake version bump we created earlier.
1. Checkout `master` locally and revert the fake version bump commit by running `git reset --hard HEAD~1`

<!-- End testing instructions -->

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

markbiek committed:
Command to tag the current release

Once we have a successful build, we create a tag for the release
version, add the build directory to it, and push it to origin.

markbiek committed:
Refactoring build command

Now we prompt to create the new release as part of the build command. We
also switch back to the previous branch if the user cancels the release
build.

andres-blanco commented on scripts/commands/build.js:
```diff
@@ -44,12 +49,28 @@ async function buildRelease() {
+		return false;
+	}
+
 	await execAsync( 'npm i' );
```
If the user accepted to go forward it would be helpful to output a message telling the user that the process might take some time. It scared me at first that it wasn't responding and I `ctrl+c`d

andres-blanco commented on scripts/commands/tag.js:
```diff
@@ -0,0 +1,82 @@
+
+	res = await tagExists( version );
+	if ( res ) {
+		error( `The tag ${ version } already exists. Deploy failed.` );
```
After getting this error my working copy was on the `master` branch instead of the `1101-update/tag-release-build` branch. I don't know if this process is meant to be run on a different branch than `master` but it felt messy to be on a different branch after an error.
<img width="1569" alt="image" src="https://user-images.githubusercontent.com/375980/235941629-ae942b2f-8a41-4658-bee1-a87151543e1f.png">

markbiek replied:
This is a bug! You're supposed to end up back at the original branch you started on after an error.

The issue here is that we switched to `master` as part of the build step. The way it's written now, the tagging step thinks the "original branch" _is_ `master`.

I'll update with a fix.

andres-blanco commented on scripts/commands/tag.js:
```diff
@@ -0,0 +1,82 @@
+		warning(
+			`Aborting. The release for ${ versionStr } was not published!`
+		);
+		// TODO - Should we delete the tag here?
```
yes, I think we should remove the recently created tag and add a message stating what happened.

andres-blanco reviewed (changes requested):
Thanks for the detailed test instructions. I've left some minor suggestions :)

markbiek committed:
Progress indicator when building the release

markbiek committed:
Pass currentBranchName to all commands

This makes it easier to switch back to the current branch at any point during the process. This is particularly important for tagRelease since the buildRelease function switches to master

markbiek committed:
Delete the new tag on abort

markbiek commented:
@andres-blanco I think all of those changes have been made. Thanks for the suggestions!

andres-blanco reviewed (changes requested):
I've found a subtle one. If you run `yarn create-release` and immediately press `ctrl+c` you'll be left in the `master` branch, instead of the one you started with. I don't know what's the best way to catch this. If it's to hard/complex to recover from the `ctrl+c` I think it's OK to leave that for later!

Everything else tests fine.

markbiek committed:
Remove unused dependency

markbiek commented:
> I've found a subtle one. If you run `yarn create-release` and immediately press `ctrl+c` you'll be left in the `master` branch, instead of the one you started with. I don't know what's the best way to catch this. If it's to hard/complex to recover from the `ctrl+c` I think it's OK to leave that for later!
> 
> Everything else tests fine.

I did a quick test to see if I could catch `ctrl+c` by using `process.on( 'SIGINT' )` and that didn't work.

I agree that it's something that would be nice to handle but it's OK to work on a fix later.
